# **EcoliBowtie2AlignmentAnalysis**
## Introduction
This project provides an efficient and user-friendly pipeline to analyze E. coli sequencing data. Leveraging the power of Bowtie2, SAMtools, and BCFtools, it aims to offer researchers and bioinformatics enthusiasts a streamlined method to align reads and detect genomic variants.

## Features:
Index building from the E. coli genome using bowtie2-build.
Efficient alignment of sequencing reads with Bowtie2.
Generation of BAM and sorted BAM files with SAMtools.
Variant calling with BCFtools to identify indels and other genomic variants.
## Prerequisites:
[Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)
[SAMtools](http://www.htslib.org/)
[BCFtools](http://www.htslib.org/doc/bcftools.html)


## Dependencies:
Before running the pipeline, ensure you have loaded the required modules on your system. You can do this by using the following commands:
```
$ module load miniconda
$ module load samtools
$ module load bcftools
$ module load bowtie2
```
> Note: The availability of these modules may depend on your computing environment. If you encounter issues or if these modules are not available, consider installing them using conda or another package manager.


## **Usage:**

> 
>#### In the sections below, I will provide step-by-step instructions for running the EcoliBowtie2AlignmentAnalysis pipeline and also explain each command and its components.
> 

### 1. Clone the Repository to your local machine:

```
git clone https://github.com/YourUsername/EcoliBowtie2AlignmentAnalysis.git
```

Next, navigate into the directory using:
```
$cd EcoliBowtie2AlignmentAnalysis
```


### 2. Index Building:
This step creates an index for a reference genome. 

```
$bowtie2-build ecolicompletegenome.fasta ecoli_index
```

>`ecolicompletegenome.fasta` is the FASTA file used containing the E. coli genome sequence.

>`ecoli_index`: This is the prefix for the index files that will be generated. 


After this command, you'll have multiple files starting with "ecoli_index" and ending with various extensions (e.g., .bt2).


### 3. Run the Alignment:

This step aligns a set of query sequences to the indexed genome.

```
$bowtie2 -x ecoli_index -f query_seqs.fasta -S output.sam
```


>`-x ecoli_index` specifies the prefix of the index files (created in the previous step).
>
>`-f` indicates that the input sequences are in FASTA format. (Use -U for unpaired FASTQ files.)
>
>`query_seqs.fasta`: This is the example file containing the query sequences to align to the E. coli genome.
>
>`-S output.sam` outputs the alignment file in SAM format.


### 4. Convert SAM to BAM and Sort:

This step converts the SAM formatted alignment to a binary BAM format.


```
$samtools view -bS output.sam > output.bam
```


>`samtools view` is used to view and convert sequence alignment files.
>
>`-b` indicates that the output should be in BAM format.
>
>`-S` indicates that the input is in SAM format.
>
>The output redirects `output.sam` SAM formatted file generated by Bowtie 2 to a file named `output.bam`.

```
$samtools sort output.bam -o output_sorted.bam
```
>`samtools sort` sorts alignments by leftmost coordinates (i.e., by their position on the reference genome).
>
>This steps takes `output.bam` (the input BAM file that you want to sort) and save the sorted alignments to a file named `output_sorted.bam`.

### 5. Variant Calling:

This steps uses two tools from the bcftools suite -- mpileup and call -- to detect variants (like SNPs and indels) from a sorted BAM file of sequence alignments. 

First, `bcftools mpileup` is used to help in visualizing how reads from the query align to each position in the ecoli reference genome. Then, we pass the pileup file in BCF format to `bcftools call` to determine where the sample differs from the reference based on the pileup of aligned reads.

```
bcftools mpileup -Ou -f ecolicompletegenome.fasta output_sorted.bam | bcftools call -Ov -mv -o variants.vcf
```

>`bcftools mpileup` generates a "pileup" (a summary of how reads align) of read bases for every position in the reference genome.
>
>`-Ou`: specifies output in the uncompressed BCF format. This format is more space-efficient and faster to process than Variant Call Format (VCF).
>
>`-f ecolicompletegenome.fasta` specifies the reference genome against which the alignments were done.
>
>`output_sorted.bam`: The sorted BAM file containing the alignments.

>`bcftools call` processes the pileup to determine which positions in the reference genome have variants based on the pileup.
>
> `-Ov` outputs in VCF format.
>
> `-mv`: calls SNP variants.
>
> `-o variants.vcf`: saves the called variants to a vcf file named `variants.vcf`

### Advanced Configurations:

> Custom Genome: If you have a custom E. coli genome, replace ecolicompletegenome.fasta with your FASTA file.
>
> Parallel Processing: To speed up the alignment, use the --threads option with Bowtie2.

### Results:
After executing the pipeline, you'll obtain a VCF file (variants.vcf) containing potential genomic variants in the analyzed reads.

### Troubleshooting:
> Alignment Errors: Ensure the index and query files are in the same directory.
>
> Dependencies: Check if all the prerequisites and dependencies are correctly installed.

## Contributing:
Feel free to fork the project, make your improvements, and submit a pull request. Your contributions are always welcome!

## License:
This project is licensed under the MIT License. Refer to the LICENSE file for more details.

## Acknowledgements:
Thanks to the BF527 course at Boston University for inspiring this project and providing the datasets.
